generic_config_updater/generic_updater.py

```diff
@@ -0,0 +1,469 @@
+            elif not tokens[0]: # Not empty
+                raise Exception("Table name in patch cannot be empty")
+            else:
+                config_db[tokens[0]] = {}
+
+        cropped_config_db = self.config_wrapper.crop_tables_without_yang(config_db)
+
+        # valid if no tables dropped during cropping
+        return len(cropped_config_db.keys()) == len(config_db.keys())
+
+    def verify_same_json(self, expected, actual):
+        # patch will be [] if no diff, [] evaluates to False
+        return not jsonpatch.make_patch(expected, actual)
+
+    def generate_patch(self, current, target):
+        return jsonpatch.make_patch(current, target)
+
+    def simulate_patch(self, patch, jsonconfig):
+        return patch.apply(jsonconfig)
+
+    def convert_config_db_patch_to_sonic_yang_patch(self, patch):
+        if not(self.validate_config_db_patch(patch)):
+            raise Exception(f"Given patch is not valid")
+
+        current_config_db = self.config_wrapper.get_config_db_as_json()
+        target_config_db = self.simulate_patch(patch, current_config_db)
+
+        current_yang = self.config_wrapper.convert_config_db_to_sonic_yang(current_config_db)
+        target_yang = self.config_wrapper.convert_config_db_to_sonic_yang(target_config_db)
+
+        return self.generate_patch(current_yang, target_yang)
+
+class ConfigFormat(Enum):
+    SONICYANG = 1
+    CONFIGDB = 2
+
+class PatchApplier:
+    def __init__( \
+        self, \
+            patchorderer = PatchOrderer(), \
+                changeapplier = ChangeApplier(), \
+                    config_wrapper = ConfigWrapper(), \
+                        patch_wrapper = PatchWrapper()):
+        self.patchorderer = patchorderer
+        self.changeapplier = changeapplier
+        self.config_wrapper = config_wrapper
+        self.patch_wrapper = patch_wrapper
+
+    def apply(self, patch):
+        # 1. Get old config as SONiC Yang
+        old_config = self.config_wrapper.get_sonic_yang_as_json()
+
+        # 2. Generate target config
+        target_config = self.patch_wrapper.simulate_patch(patch, old_config)
+
+        # 3. Validate target config
+        if not(self.config_wrapper.validate_sonic_yang_config(target_config)):
+            raise Exception(f"The given patch is not valid")
+
+        # 4. Generate list of changes to apply
+        changes = self.patchorderer.order(patch)
+
+        # 5. Apply changes in order
+        for change in changes:
+            self.changeapplier.apply(change)
+
+        # 6. Validate config updated successfully
+        new_config = self.config_wrapper.get_sonic_yang_as_json()
+        if not(self.patch_wrapper.verify_same_json(target_config, new_config)):
+            raise Exception(f"After applying patch to config, there are still some parts not updated")
+
+class ConfigReplacer:
+    def __init__(self, patch_applier = PatchApplier(), config_wrapper = ConfigWrapper(), patch_wrapper = PatchWrapper()):
```

Comment:
>PatchApplier() [](http://example.com/codeflow?start=39&length=14)

Do not use complex mutable object as default value. ref: [https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html](https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html)